### PR TITLE
Conditionally enable Groovy 4 for testing

### PR DESCRIPTION
**Pull request by lptr**
This is a cleaned up version of https://github.com/gradle/gradle/pull/18552.

Building with `-DbundleGroovy4=true` on the CLI will use Groovy 4.0.7 libs.  An additional Gradleception job is already in place to invoke this code path.

If we choose to unconditionally upgrade to Groovy 4, we should revert this PR followed by setting the Groovy version value in `ExternalModulesExtension.kt`.

Alternatively, we could modify this PR to test forward with Groovy 5.  This would likely also need temporary inclusion of apache snapshot repositories; see #20972. 

## Preparation PRs

- https://github.com/gradle/gradle/pull/18665
- https://github.com/gradle/gradle/pull/20037
- #21504
- #21497
- #21986
- #22097
- #22130
- #22133 
- #22151
- #22157
- #22158
- #22178
- #22208 
- #22213 

## TODO

- [x] allow configuration-cache instrumentation to handle `INVOKEDYNAMIC`
  - https://github.com/gradle/gradle/pull/20795
  - https://github.com/gradle/gradle/pull/20799
  - https://github.com/gradle/gradle/pull/20616
- [x] fix capabilities
- [ ] mention relevant breaking changes in release notes/upgrade guide
  - [x] `is*()` getters are not supported for `Boolean` properties anymore; see #22218
  - [ ] `INVOKEDYNAMIC` is the only way forward
- [x] upgrade Spock to Groovy 4-compatible version – https://github.com/spockframework/spock/pull/1382
  - [x] upgrade to Spock 2.2 GA
  - [x] upgrade to Spock 2.2-M1
- [x] upgrade CodeNarc – https://github.com/gradle/gradle/pull/20655
  - [x] https://github.com/CodeNarc/CodeNarc/issues/678
    - this prevents running CodeNarc because it depends on the now-removed `groovy.utils.XmlParser`
    - it makes `IsolatedAntBuilderMemoryLeakIntegrationTest.CodeNarc does not fail with PermGen space error` fail
  - [x] https://github.com/CodeNarc/CodeNarc/issues/682
- [x] upgrade to Groovy 4
  - [x] upgrade to Groovy 4.0.7
  - [x] [GROOVY-10765](https://issues.apache.org/jira/browse/GROOVY-10765) _STC: Closure implementation of Java @FunctionalInterface loses type information_
    - Can workaround using a cast in UnitOfWorkBuilder
    - Fix coming in 4.0.6
  - [x] [GROOVY-10731](https://issues.apache.org/jira/browse/GROOVY-10731) _Exceptions thrown from MarkupTemplateEngine when map accessors and GString interpolation are used_
    - fixed in 4.0.5
  - [ ] [GROOVY-10709](https://issues.apache.org/jira/browse/GROOVY-10709) _Performance regression in Gradle with Groovy 4_
    - no clear cause; mitigation/investigation is ongoing
  - [x] [GROOVY-10708](https://issues.apache.org/jira/browse/GROOVY-10708) _Property access for wrapped vs. primitive boolean_
    - Groovy team will not fix
    - @big-guy has suggested a workaround in Gradle to prevent breaking legacy plugins/tasks, to be implemented as #22218
  - [x] [GROOVY-10772](https://issues.apache.org/jira/browse/GROOVY-10772) _Possible memory leak, CacheableCallSite retains objects across invocations_
    - Causes OOME-related flakiness in `o.g.w.i.WorkerExecutorIntegrationTest#does not leak project state across multiple builds`, and OOME mostly happens with Java 8
  - [x] [GROOVY-10707](https://issues.apache.org/jira/browse/GROOVY-10707) _Regression in property access for conflicting isXxx and getXxx accessors_
    - Groovy team will not fix
    - Our tests are refactored to accommodate the breaking change 
  - [x] [GROOVY-10591](https://issues.apache.org/jira/browse/GROOVY-10591) _Static method not found on Java interface_
    - no fix planned yet
  - [x] [GROOVY-10543](https://issues.apache.org/jira/browse/GROOVY-10543) _Problem with groovy-all Gradle module metadata_
    - fixed in 4.0.2
  - [x] [GROOVY-10514](https://issues.apache.org/jira/browse/GROOVY-10514) _Class method is not called anymore from inside closure_
    - apparently works as designed, added workarounds
  - [x] [GROOVY-10555](https://issues.apache.org/jira/browse/GROOVY-10555) _MetaClass.getProperites() returns package private fields_
    - no plan to fix yet, but added a workaround
  - [x] [GROOVY-10521](https://issues.apache.org/jira/browse/GROOVY-10521) _Compiler complains about abstract method not implemented when implementing trait_
    - not fixed yet, but we added a workaround
  - [x] ~~[GROOVY-10466](https://issues.apache.org/jira/browse/GROOVY-10466) _Compilation error on Spock expectation_~~
    - To be fixed in https://github.com/apache/groovy/pull/1698
    - Fixed in 4.0.2-SNAPSHOT
  - [x] ~~[GROOVY-10467](https://issues.apache.org/jira/browse/GROOVY-10467) _Compilation fails with method detected as transient_~~
    - doesn't seem to be a problem with `4.0.1`
  - [x] ~~[GROOVY-10299](https://issues.apache.org/jira/browse/GROOVY-10299) _Groovy compiler generates invalid Java stubs_~~
    - already fixed in `4.0.0-beta-2`
  - [x] ~~[GROOVY-10290](https://issues.apache.org/jira/browse/GROOVY-10290) _Dynamic Groovy code in Gradle doesn't compile because of $getLookup() method is not static_~~
    - fixed in `4.0.1`
- [x] fix build scripts being compiled with binary compatibility (=class format version) of the runtime Java version
  - this makes the following tests fail:
    - `WorkerDaemonIntegrationTest`
    - `ConfigureRuntimeClasspathNormalizationIntegrationTest`
    - `CheckstylePluginToolchainsIntegrationTest`
    - `BuildEnvironmentIntegrationTest`
- [x] use non-deprecated Groovy `AntBuilder` for the super-class of our own `AntBuilder`
- [x] upgrade Groovy version in `AbstractSourcesAndJavadocJarsIntegrationTest`
- [x] use `org.apache.groovy` group ID for Groovy versions > 4 in `GroovyParametersMetadataIntegrationTest` (and probably in many other places)

### CODE REVIEW

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please

**Comment by lptr** [comment]
@bot-gradle test this please

**Comment by lptr** [comment]
@bot-gradle run performance tests.

**Comment by lptr** [comment]
@bot-gradle test AllPerformanceTest please.

**Comment by lptr** [comment]
@bot-gradle test this please

**Comment by lptr** [comment]
@bot-gradle test this please

**Comment by lptr** [comment]
@bot-gradle test this please

**Comment by lptr** [comment]
@bot-gradle test this please

**Comment by lptr** [comment]
@bot-gradle test this please

**Comment by lptr** [comment]
@bot-gradle test this please

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
@bot-gradle test QF

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
@bot-gradle test QF

File: build-logic/basics/src/main/kotlin/gradlebuild.repositories.gradle.kts [hunk 1]
```diff
@@ -39,5 +39,12 @@ repositories {
             includeGroup("io.usethesource")
         }
     }
+    maven {
+        name = "Apache Snapshots"
+        url = uri("https://repository.apache.org/content/repositories/snapshots/")
+        content {
+            includeGroup("org.apache.groovy")
+        }
+    }
```
Comment (DPUkyle) [inline build-logic/basics/src/main/kotlin/gradlebuild.repositories.gradle.kts:48]: This was probably useful earlier, but I vote we remove this before merging.

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
@bot-gradle test QF

**Comment by DPUkyle** [comment]
@bot-gradle help

**Comment by DPUkyle** [comment]
@bot-gradle cancel

**Comment by DPUkyle** [comment]
@bot-gradle test QF

**Comment by DPUkyle** [comment]
@bot-gradle test QF

**Comment by eskatos** [comment]
@bot-gradle test APT

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/JUnitMultiVersionIntegrationSpec.groovy [hunk 2]
```diff
@@ -38,7 +40,7 @@ import static org.gradle.test.fixtures.junitplatform.JUnitPlatformTestRewriter.*
  */
 abstract class JUnitMultiVersionIntegrationSpec extends MultiVersionIntegrationSpec {
     // JUnit 5's test case name contains parentheses which might break test assertion, e.g. testMethod() PASSED -> testMethod PASSED
-    private static final Pattern TEST_CASE_RESULT_PATTERN = ~/(.*)(\w+)\(\) (PASSED|FAILED|SKIPPED|STANDARD_OUT)/
+    static final Pattern TEST_CASE_RESULT_PATTERN = ~/(.*)(\w+)\(\) (PASSED|FAILED|SKIPPED|STANDARD_OUT)/
```
Comment (big-guy) [inline subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/JUnitMultiVersionIntegrationSpec.groovy:43]: safe?

File: subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/DefaultScriptCompilationHandler.java [hunk 3]
```diff
@@ -195,6 +195,7 @@ private void wrapCompilationFailure(ScriptSource source, MultipleCompilationErro
     private CompilerConfiguration createBaseCompilerConfiguration(Class<? extends Script> scriptBaseClass) {
         CompilerConfiguration configuration = new CompilerConfiguration();
         configuration.setScriptBaseClass(scriptBaseClass.getName());
+        configuration.setTargetBytecode(CompilerConfiguration.JDK8);
```
[the remark this answers is not in the document]
Comment (DPUkyle) [reply subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/DefaultScriptCompilationHandler.java:198]: This should be safe to port to master, as we already require Java 8 for running Gradle. Groovy 4 "catches up" to us and adds the same requirement.

File: subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/FileCacheBackedScriptClassCompiler.java [hunk 4]
```diff
@@ -412,6 +413,11 @@ public void visitMethodInsn(int opcode, String owner, String name, String desc,
                 mv.visitMethodInsn(opcode, remap(owner), name, remap(desc), intf);
             }
 
+            @Override
+            public void visitInvokeDynamicInsn(String name, String descriptor, Handle bootstrapMethodHandle, Object... bootstrapMethodArguments) {
```
[the remark this answers is not in the document]
Comment (DPUkyle) [reply subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/FileCacheBackedScriptClassCompiler.java:417]: In theory yes, though it should never be called until we include/enable groovy-indy.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/attributes/VariantAwareResolutionWithConfigurationAttributesIntegrationTest.groovy [hunk 5]
```diff
@@ -144,11 +144,12 @@ class VariantAwareResolutionWithConfigurationAttributesIntegrationTest extends A
 
     @ToBeFixedForConfigurationCache
     def "compiling project variant doesn't imply execution of other variants build tasks"() {
+        testDirectory.mkdirs()
```
[the remark this answers is not in the document]
Comment (DPUkyle) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/attributes/VariantAwareResolutionWithConfigurationAttributesIntegrationTest.groovy:147]: Smells like debugging code; I'll make a note to revert this line.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/attributes/VariantAwareResolutionWithConfigurationAttributesIntegrationTest.groovy [hunk 6]
```diff
@@ -144,11 +144,12 @@ class VariantAwareResolutionWithConfigurationAttributesIntegrationTest extends A
...
         def projectDir = new FileTreeBuilder(testDirectory)
         given:
         projectDir {
-            withVariants(buildFile)
-            withExternalDependencies(buildFile, '''
+            VariantAwareResolutionWithConfigurationAttributesIntegrationTest.withVariants(buildFile)
```
[the remark this answers is not in the document]
Comment (DPUkyle) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/attributes/VariantAwareResolutionWithConfigurationAttributesIntegrationTest.groovy:151]: See 048d449d81c282bf9369655f7de4f815e74a3362 and 563b106e35dc033c272500e7c3926536f42bdc54

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/rules/IvySpecificComponentMetadataRulesIntegrationTest.groovy [hunk 7]
```diff
@@ -55,7 +55,7 @@ task resolve {
         repository {
             'org.test:projectA:1.0' {
                 withModule {
-                    withExtraInfo((ns('foo')): "fooValue", (ns('bar')): "barValue")
+                    withExtraInfo((IvySpecificComponentMetadataRulesIntegrationTest.ns('foo')): "fooValue", (IvySpecificComponentMetadataRulesIntegrationTest.ns('bar')): "barValue")
```
[the remark this answers is not in the document]
Comment (DPUkyle) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/rules/IvySpecificComponentMetadataRulesIntegrationTest.groovy:58]: See b868d143bc3a241f2425dc814baacd5f9a2f4e14

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/resolver/DefaultGradleApiSourcesResolver.java [hunk 8]
```diff
@@ -37,7 +37,7 @@ public class DefaultGradleApiSourcesResolver implements GradleApiSourcesResolver
 
     private static final String GRADLE_LIBS_REPO_URL = "https://repo.gradle.org/gradle/list/libs-releases";
     private static final String GRADLE_LIBS_REPO_OVERRIDE_VAR = "GRADLE_LIBS_REPO_OVERRIDE";
-    private static final Pattern FILE_NAME_PATTERN = Pattern.compile("(groovy(-.+)?)-(\\d.*?).jar");
+    private static final Pattern FILE_NAME_PATTERN = Pattern.compile("(groovy(-.+?)?)-(\\d.+?)\\.jar");
```
[the remark this answers is not in the document]
Comment (DPUkyle) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/resolver/DefaultGradleApiSourcesResolver.java:40]: Probably; I speculate this had to do with enabling testing with snapshot versions.

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/AbstractBuildExperimentRunner.java [hunk 9]
```diff
@@ -133,7 +134,7 @@ private File outputDirFor(String testId, BuildExperimentSpec spec) {
             String version = ((GradleBuildExperimentSpec) spec).getInvocation().getGradleDistribution().getVersion().getVersion();
             outputDir = new File(outputDirSelector.outputDirFor(testId), version);
         } else {
-            outputDir = new File(outputDirSelector.outputDirFor(testId), OutputDirSelector.fileSafeNameFor(spec.getDisplayName()));
+            outputDir = new File(outputDirSelector.outputDirFor(testId), OutputDirSelectorUtil.fileSafeNameFor(spec.getDisplayName()));
```
[the remark this answers is not in the document]
Comment (DPUkyle) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/AbstractBuildExperimentRunner.java:137]: Will cherry-pick this change for now, with a note to investigate the CodeNarc rule later.

File: subprojects/language-groovy/src/main/java/org/gradle/api/internal/tasks/compile/ApiGroovyCompiler.java [hunk 10]
```diff
@@ -194,7 +194,7 @@ public WorkResult execute(final GroovyJavaJointCompileSpec spec) {
         // Disallow classes from Groovy Jar that reference external classes. Such classes must be loaded from astTransformClassLoader,
         // or a NoClassDefFoundError will occur. Essentially this is drawing a line between the Groovy compiler and the Groovy
         // library, albeit only for selected classes that run a high risk of being statically referenced from a transform.
-        groovyCompilerClassLoaderSpec.disallowClass("groovy.util.GroovyTestCase");
+        groovyCompilerClassLoaderSpec.disallowClass("groovy.test.GroovyTestCase");
```
[the remark this answers is not in the document]
Comment (DPUkyle) [reply subprojects/language-groovy/src/main/java/org/gradle/api/internal/tasks/compile/ApiGroovyCompiler.java:197]: Will fix.

File: subprojects/model-core/src/test/groovy/org/gradle/api/internal/provider/CollectionPropertySpec.groovy [hunk 11]
```diff
@@ -617,7 +617,7 @@ The value of this property is derived from: <source>""")
         def value = property.calculateExecutionTimeValue()
         value.isFixedValue()
         !value.hasChangingContent()
-        value.fixedValue.isEmpty()
```
[the remark this answers is not in the document]
Comment (DPUkyle) [reply subprojects/model-core/src/test/groovy/org/gradle/api/internal/provider/CollectionPropertySpec.groovy:620]: I agree, having two accessors that appear to represent the same property is ugly.

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/JUnitMultiVersionIntegrationSpec.groovy (hunk 2 above)
Comment (DPUkyle) [reply subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/JUnitMultiVersionIntegrationSpec.groovy:43]: This seems unnecessary - I find no usages outside this class.

File: subprojects/plugins/src/integTest/groovy/org/gradle/groovy/compile/BasicGroovyCompilerIntegrationSpec.groovy [hunk 12]
```diff
@@ -57,9 +57,6 @@ abstract class BasicGroovyCompilerIntegrationSpec extends MultiVersionIntegratio
 
     def "compileGoodCode"() {
         if (module == "groovy-all") {
-            // Do not test with groovy-all with Groovy 4 for now because it doesn't work as a platform currently
-            // See https://issues.apache.org/jira/browse/GROOVY-10543
-            Assume.assumeTrue(versionNumber.major < 4)
```
[the remark this answers is not in the document]
Comment (DPUkyle) [reply subprojects/plugins/src/integTest/groovy/org/gradle/groovy/compile/BasicGroovyCompilerIntegrationSpec.groovy:62]: This one confused me a bit, let's discuss.

File: subprojects/model-core/src/testFixtures/groovy/org/gradle/api/internal/provider/ProviderAssertions.groovy [hunk 13]
```diff
@@ -33,9 +33,20 @@ trait ProviderAssertions {
         producer.visitContentProducerTasks { assert false }
     }
 
-    void assertHasProducer(ProviderInternal<?> provider, Task task, Task... additional) {
-        def expected = [task] + (additional as List)
+    // Varargs blown up because of Groovy compiler issue: https://issues.apache.org/jira/browse/GROOVY-10521
```
[the remark this answers is not in the document]
Comment (DPUkyle) [reply subprojects/model-core/src/testFixtures/groovy/org/gradle/api/internal/provider/ProviderAssertions.groovy:36]: I can confirm this is not yet fixed: https://issues.apache.org/jira/browse/GROOVY-10521?focusedCommentId=17581565&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17581565

**Review by nathan-contino — APPROVED** [review]
Docs LGTM

**Review by big-guy — APPROVED** [review]

**Comment by DPUkyle** [comment]
@bot-gradle test and merge

**Comment by DPUkyle** [comment]
@bot-gradle test RFN

**Comment by DPUkyle** [comment]
Note: this is currently blocked pending resolution of [GROOVY-10772](https://issues.apache.org/jira/browse/GROOVY-10772)

**Comment by lptr** [comment]
Hm. When looking at performance regressions I remember that in some of our tests we've seen an increase in GC time. Could this be the reason, I wonder?

**Review by hythloda — APPROVED** [review]

**Comment by octylFractal** [comment]
@bot-gradle test PT

**Comment by big-guy** [comment]
@bot-gradle test this

**Review by big-guy — APPROVED** [review]

**Comment by octylFractal** [comment]
@bot-gradle test and merge

**Pull request merged**